### DISCUSSION TITLE
rdk plugin - dots move in opposite direction to specified (issue #922)

### DISCUSSION THREAD

**Discussion by becky-gilbert** [jspsych/jsPsych#932] [Q&A] [answered]
This is an issue (#922) with the rdk plugin reported by @majorminors. It does look like a bug to me but I'm not entirely sure. So I'm posting it here to see if we can get some help with it. 

In Dorian's task, there are two possible dot motion directions on each trial. When the two possible directions are 45 degrees and 225 degrees, the dots move in the expected directions. However, when the two possible directions are 135 or 315 degrees, the dots appear to move in the exact _opposite_ direction than specified in the code. Dorian has created a [minimal example](https://github.com/majorminors/evaccum-js/blob/coh_dir_bug/coh_issue_test.html), which you can download/clone and run to see the problem.

I got as far as finding the functions in the RDK plugin code that calculate the [change in X coordinate](https://github.com/jspsych/jsPsych/blob/83980085ef604c815f0d97ab55c816219e969b84/plugins/jspsych-rdk.js#L743) and [change in Y coordinate](https://github.com/jspsych/jsPsych/blob/83980085ef604c815f0d97ab55c816219e969b84/plugins/jspsych-rdk.js#L749) for each coherent dot, and the function where these change values are [saved into the dot object](https://github.com/jspsych/jsPsych/blob/83980085ef604c815f0d97ab55c816219e969b84/plugins/jspsych-rdk.js#L1119), and then [added on to the dot's existing X and Y coordinates](https://github.com/jspsych/jsPsych/blob/83980085ef604c815f0d97ab55c816219e969b84/plugins/jspsych-rdk.js#L1136). If there is a bug, then my guess it's related to how the change in dot position values are calculated and transformed for the canvas' coordinate system, which happens in the `calculateCoherentJumpSizeX` and `calculateCoherentJumpSizeY` functions. However I'm not sure whether or not these functions are written correctly because my math is not up to scratch!

It would be helpful to get confirmation from others that this is bug. And if it is a bug, we'd be very interested in suggestions for fixing it. Thanks!

**Comment by jodeleeuw** [comment]
Tagging in @vrsivananda who is likely to be fastest to ID if this is a bug.

**Comment by vrsivananda** [comment, marked as answer]
Thanks @jodeleeuw for alerting me to this issue.

And thanks @becky-gilbert and @majorminors for the minimal example code as well. I ran it and it seems to me to be working as intended, in that the dots are moving in the directions as specified in the minimal example. It starts off at 45 degrees and then for each of the subsequent 3 trials, it rotates in 90-degree increments in a counter-clockwise direction. 

So I think there isn't a bug, or I might be misunderstanding your example.

**Reply by becky-gilbert** [reply]
Hi @vrsivananda, thanks so much for having a look at this! 

You're right that in the example the dots should be moving at 45, 135, 225, and 315 across the 4 trials. However, when both @majorminors and I run this code, we see the dots moving in the opposite directions for the 2nd and 4th trials, i.e. 135 and 315 degrees. In order to see this more clearly, I suggest increasing the `dot_life` parameter to 40 and `trial_duration` to 4000. I'll create a clearer example when I get a chance.

Also, I checked the output of the functions that calculate the change in dot coordinates, `calculateCoherentJumpSizeX` and `calculateCoherentJumpSizeY`. Here's the pattern that I get in terms of the signs of the coordinate changes for each coherent direction:

45 deg: X+, Y-
**135 deg: X-, Y-**
225 deg: X-, Y+
**315 deg: X+, Y+**

These coordinate changes seem correct for 45 and 225 degrees. E.g. for 45 deg, the increase in X moves the dot toward the right, and the decrease in Y moves it up.

But for 135 degrees, when the dots should be moving toward the lower-right corner of the HTML canvas, the change values are both _negative_, which moves the dot toward the upper-left corner, i.e. toward (0,0). And a similar issue happens with 315 degrees - the changes values are both _positive_, which moves the dot toward the lower-right corner on the HTML canvas, rather than the upper-left.

Again, we're not certain that this is a bug, but we can't figure out why we're not getting the desired behavior.

**Reply by vrsivananda** [reply]
Thanks @becky-gilbert  for the clear explanation. I think the reason why it seems incorrect is because the direction starts at **0 degrees** (which moves it **towards the right**) and then the increment in direction is **counterclockwise**. This might just be a misunderstanding. I infer from your explanation that you expected 0 degrees to be moving upwards and then going clockwise from there.

Here is a snippet from the [documentation](https://www.jspsych.org/plugins/jspsych-rdk/) of the `coherent_direction` parameter:

> The direction of movement for coherent dots in degrees. 0 degrees is in the 3 o'clock direction, and increasing this number moves counterclockwise. (E.g. 12 o'clock is 90, 9 o'clock is 180, etc.) Range is 0 - 360.

So just to be sure we're on the same page, this is the expected behavior:
```
 45 degrees --> Upper right --> X+, Y-
135 degrees --> Upper left  --> X-, Y-
225 degrees --> Lower left  --> X-, Y+
315 degrees --> Lower right --> X+, Y+
```

Let me know if I am still missing something.

**Reply by becky-gilbert** [reply]
Aha! Thank you, yes that's the piece we were missing. Sorry to have missed this in the documentation, and thanks very much for the explanation!

**Reply by vrsivananda** [reply]
No worries, I'm glad this is resolved!

**Reply by majorminors** [reply]
Missed all of this! Many thanks for the help, both @becky-gilbert  and @vrsivananda! Apologies for missing it in the docs.
